allow sending tracing metadata (target, file/line, module_path) as syslog structured data.

cbeck88 opened the pull request:
this adds optional features to pass on more of the metadata from the tracing infrastructure as structured-data, when using RFC-5424 format.

it doesn't go as far as sending fields and valuesets from tracing events, but still seems like a significant improvement.

note that `tracing_subscriber` typically defaults to formatting the target with the log message in text logs, so many users are used to seeing that in text logs, but it's not present in the syslog records produced by this crate before this change. in tracing_subscriber, the file and line number are optional defaulting to off. in this PR we make all this stuff default to off, for backwards compatible behavior.

cbeck88 committed:
allow sending tracing metadata (target, file/line, module_path) as syslog structured data

this adds optional features to pass on more of the metadata from the tracing infrastructure
as structured-data, when using RFC-5424 format.

it doesn't go as far as sending fields and valuesets from tracing events, but still seems
like a significant improvement.

note that `tracing_subscriber` typically defaults to formatting the target
with the log message in text logs, so many users are used to seeing that in text logs,
but it's not present in the syslog records produced by this crate before this change.
in tracing_subscriber, the file and line number are optional defaulting to off.
in this PR we make all this stuff default to off, for backwards compatible behavior.

sp1ff commented:
This seems fine; my only concern is the unconditional use of the `meta` SD-ID (AKA Structured Data IDentifier). Per the [RFC](https://datatracker.ietf.org/doc/html/rfc5424#section-7.3), there are only three parameters that can be used in this structured data item: `sequenceId`, `sysUptime`, and `language`. Putting `tracing` metadata in here is non-conformant.

Now, if the caller *wants* to do that, we certainly shouldn't stop them, but I don't like forcing them. They may have a Private Enterprise Number, for instance, allowing them to include this information in a compliant manner. I'm also considering applying for a PEN, to be associated with this crate, for this purpose-- what do you think?

cbeck88 commented:
interesting, I think I didn't read this part of the RFC carefully :sweat_smile:

it sounds like we could allow the user to supply an SD-ID for this -- if for example they already have a PEN?

applying for a PEN for the crate sounds nice :+1:

sp1ff commented:
Alright, I've applied for a PEN (status can be checked [here](https://www.iana.org/requests/pgxa-20s-3vn/)). If it's not going to cause you too much trouble, can we put this PR on hold until IANA acts on this? If I get one, I'll make a sub-assignment to this crate for this purpose, and you can update the PR to use that by default, or use a caller-supplied SD-ID.

sp1ff commented:
I have a PEN: 64700 (see [here](https://www.iana.org/assignments/enterprise-numbers/enterprise-numbers)). I propose we use the SD-ID "tracing-meta@64700" by default, but allow the user to override that. Thoughts?

Also, is addition of the metadata to the formatted message configurable (like, with a configuration bit in the `Formatter` implementation)?

cbeck88 commented:
this is great! i will update my PR

cbeck88 committed:
add optional integration with tracing-log

tracing-log is a bridge that allows tracing to pick up log messages
written by older rust libraries that used the `log` crate facade.
the bridge subscribes to log messages and resubmits them as tracing
events.

because of lifetime issues, it can't take the log message file and
line numbers and register them directly as file and line numbers of
the tracing events. instead it places them as `log.file` and `log.line`
in the event metadata. then, when using `tracing-subscriber`, the
metadata are "normalized" by repopulating the file and line number
information. This normalization is possible when `tracing-log` feature
is turned on.

in this commit we mirror this behavior, so that tracing-rfc-5424 will
also make log messages carry this metadata just as well as tracing messages,
if the optional `tracing-log` feature is turned on

cbeck88 commented:
i had a strange bug which i finally debugged in my application around this feature. it turns out that when you have some libraries using the `log` crate for logging, and you use the `tracing-log` bridge to map those to tracing events, their metadata is a little wonky and needs to be "normalized". see https://github.com/sp1ff/syslog-tracing/pull/15/commits/6a0b930f60ef97a81c1f11d8f107298f275df2d3 and lmk if you agree with this or have any questions about it

cbeck88 committed:
use "tracing-meta@64700" as the SD-ID by default, and allow override

this also refactored the tracing-metadata related options to be one
structure, maybe a little tidier

cbeck88 commented:
> I have a PEN: 64700 (see [here](https://www.iana.org/assignments/enterprise-numbers/enterprise-numbers)). I propose we use the SD-ID "tracing-meta@64700" by default, but allow the user to override that. Thoughts?
> 
> Also, is addition of the metadata to the formatted message configurable (like, with a configuration bit in the `Formatter` implementation)?

the configuration is now this optional substructure, which is None by default:

```
/// A syslog formatter that produces RFC [5424]-conformant syslog messages.
///
/// [5424]: https://datatracker.ietf.org/doc/html/rfc5424
pub struct Rfc5424 {
    facility: Facility,
    hostname: Hostname,
    appname: AppName,
    pid: ProcId,
    with_bom: bool,
    with_tracing_metadata: Option<TracingMetadata>,
}

#[derive(Default)]
struct TracingMetadata {
    sd_id: String,
    target: bool,
    module: bool,
    source_location: bool,
}
```

hopefully this is tidier.

the idea is that if the user sets any of these tracing_metadata things using the builder, then this struct is populated with default values. when we go to log, if it's present then we include the tracing_metadata structured data. if the sd_id is empty, then we use the default "tracing-meta@64700"

cbeck88 committed:
fix tests

cbeck88 commented:
i think i addressed everything, lmk what you think

cbeck88 commented on tracing-rfc-5424/Cargo.toml:
```diff
@@ -20,4 +26,8 @@ hostname = "0.3.1"
 tracing = "0.1.35"
 tracing-core = "0.1.28"
+tracing-log = { version = "0.2", optional = true }
 tracing-subscriber = "0.3.15"
```
Actually, looking at this again, tracing-subscriber by default depends on tracing-log.

https://docs.rs/crate/tracing-subscriber/latest/features

So there is no reason that we should make tracing-log an optional dep here, unless we also want to make it optional when we configure tracing subscriber.

I'll let you decide which direction we should go in.

sp1ff commented:
Added a period to the description to kick-off a CI run.

Ah, apologies: CI was disabled when you cut your PR and I think that kept it from running. Are you using the nightly toolchain, by any chance.

I've downloaded this locally, and as I start to dig into it, I'm realizing this is a big change-- thanks for doing this. To help me, and to update the `NEWS` file, can you summarize the change? I'm seeing adding `tracing` metadata as structured data in the syslog event, behind a feature gate. But what's this about a bridge from tracing-log?

cbeck88 commented:
> Added a period to the description to kick-off a CI run.
> 
> Ah, apologies: CI was disabled when you cut your PR and I think that kept it from running. Are you using the nightly toolchain, by any chance.
> 
> I've downloaded this locally, and as I start to dig into it, I'm realizing this is a big change-- thanks for doing this. To help me, and to update the `NEWS` file, can you summarize the change? I'm seeing adding `tracing` metadata as structured data in the syslog event, behind a feature gate. But what's this about a bridge from tracing-log?

i don't think i'm using the nightly toolchain, but i think maybe `Option::get_or_insert_default` was not stabilized until this past year? the docu says that was rust 1.83, but you are pinning to 1.82 in CI, so I guess that's the reason? I can use `get_or_insert_with(Default::default)` instead

about the tracing-log:

* `log` crate was the original log facade that everyone used, way before async.
* then async happened, and the tokio people created `tracing`.
* to try to help adoption of `tracing`, they wanted a way that any messages sent to `log` could be seamlessly consumed by the `tracing-subscriber`. that machinery is in a crate called `tracing-log`. it is included and turned on by default whenever you use the `tracing-subscriber` crate with default features (as you do in `tracing-rfc-5424`)
* they have to hack around some API differences between the `log` facade and the `tracing` facade, so the file and line metadata get stuffed as fields or something when a log message is transcribed to tracing. then, the tracing subscriber uses this `NormalizeMetadata` machinery to turn it back into a normal tracing crate `Metadata` object.
* in order for this crate to work well when `tracing-log` is used, it also needs to know to normalize the metadata. i added code that does that, which is dependent on the `tracing-log` feature, because I assumed that this should be optional behavior. however, i have since learned (by reading further) that `tracing-log` is already a non-optional, indirect dependency of this crate, because you use `tracing-subscriber` with default features. so now i'm questioning why i'm making it an explicit, optional dependency, in this revision
* I cared about `tracing-log` because I have dependencies in my app that are based on the `log` facade and not the `tracing` facade, and will probably never migrate. And so if I don't put this `tracing_log::NormalizeMetadata` stuff in `tracing-rfc-5424`, then it fails to find their file and line number information, which then doesn't make into the syslog messages. Since this is exactly what `tracing_subscriber` does internally, it seems like the right thing to do in this crate as well.

there are two directions we can go in:
* `tracing-log` becomes a mandatory dependency of this crate, and some of the cfg stuff goes away -- we will unconditionally call "normalize metadata" etc., and `tracing-rfc-5424` will not be usable without `tracing-log`.
* `tracing-log` becomes an optional dependency of this crate. we keep the cfg stuff. when `tracing-rfc-5424` depends on `tracing-subscriber`, we turn off default-features, and if `tracing-log` is on here, it gets forwarded to `tracing-subscriber`. this way users who don't need `tracing-log` can turn that feature off and it won't get built.

The first option is less work for you. However, if you'd like to make it easier for users of this crate to be able to turn off tracing-log, then it's probably better to go with the second option. We basically did all the work to support that in the current revision of this PR. It's really up to you and what you think the scope of the crate should be, whether it's interesting to you to expose more of the optional features of `tracing-subscriber` as optional features of your crate, or if you have little interest in that development direction.

cbeck88 committed:
cargo fmt

cbeck88 committed:
respect MSRV of 1.82

cbeck88 commented:
rebased on latest release

cbeck88 committed:
fix tests

cbeck88 commented:
Here's a summary

```
* Support for sending tracing metadata (target, file/line, module_path) as syslog structured data when using RFC 5424. This can be enabled via the new APIs on the RFC 5424 builder. When enabled, outbound messages will include e.g. `[tracing-metadata@64700 file="foo.rs" line="97"]`, to be consumed downstream. The SD-ID used here can also be customized.
```

Depending on if you want `tracing-log` to be optional, there's a second news item:

```
* Many unused features of `tracing-subscriber` were turned off. Additionally, `tracing-log` is now an optional dependency of `tracing-rfc5424` rather than a mandatory one, but is on by default for backwards compatibility.
```

cbeck88 commented:
i'll keep developing this on the assumption that `tracing-log` should become optional, since that's the point we are closest to. if you decide you don't want that, just lmk and we'll simplify after.

cbeck88 committed:
fixup configuration of tracing-subscriber crate to only enable required features

make tracing-log on by default for backwards compat

sp1ff commented:
This looks awesome; thanks! Let me futz around with the CI (still not sure why it wasn't running on this PR), then I'll merge & cut a release.

cbeck88 commented on tracing-rfc-5424/src/rfc5424.rs:
```diff
@@ -423,26 +679,3 @@ impl SyslogFormatter for Rfc5424 {
-    use super::*;
-
-    #[test]
-    fn test_against_issue_014_regression() {
```
this test got moved to layer.rs, because after the change, it needs a CALLSITE in order to invoke this function, but all the machinery for that was in layer.rs . if you don't like this, happy to rearrange it

cbeck88 commented on tracing-rfc-5424/src/layer.rs:
```diff
@@ -238,6 +245,18 @@ where
+        // file/line info for events that originated from the `log` crate.
+        // For native tracing events, normalized_metadata() returns None and we use
+        // the event's own metadata.
+        // See: https://github.com/tokio-rs/tracing/blob/master/tracing-subscriber/src/fmt/fmt_layer.rs
```
this should probably be a permalink

cbeck88 commented:
i'm seeing this error in CI

> 
  Downloaded syslog_rfc5424 v0.10.0
error: failed to parse manifest at `/home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/syslog_rfc5424-0.10.0/Cargo.toml`

Caused by:
  feature `edition2024` is required

  The package requires the Cargo feature called `edition2024`, but that feature is not stabilized in this version of Cargo (1.82.0 (8f40fc59f 2024-08-21)).
  Consider trying a newer version of Cargo (this may require the nightly release).

this means that we can't build the parser crate i was using in tests because we've pinned to a version of rust that is too old to use edition 2024. lmk what you want to do here, i can remove these tests i guess or maybe they can use conditional compilation. or we can use a different parser i guess

sp1ff commented:
Yeah, I should definitely upgrade to the 2024 edition, but I'd like to do that in a separate patch.

The CI fix is trivial. I've made it locally, but I don't have permission to patch to your repo. Here's the diff:
```
diff --git a/test-rfc-5424/Cargo.toml b/test-rfc-5424/Cargo.toml
index a291dc8..1ee9dc3 100644
--- a/test-rfc-5424/Cargo.toml
+++ b/test-rfc-5424/Cargo.toml
@@ -4,6 +4,7 @@ version = "0.1.5"
 edition = "2021"

 [dependencies]
+syslog_rfc5424 = "0.9"
 tracing = "0.1.36"
 tracing-rfc-5424 = { version = "0.1", path = "../tracing-rfc-5424"}
 tracing-subscriber = "0.3.15"
diff --git a/tracing-rfc-5424/Cargo.toml b/tracing-rfc-5424/Cargo.toml
index 216877b..3379060 100644
--- a/tracing-rfc-5424/Cargo.toml
+++ b/tracing-rfc-5424/Cargo.toml
@@ -30,5 +30,5 @@ tracing-log = { version = "0.2", optional = true }
 tracing-subscriber = { version = "0.3.22", default-features = false, features = ["std"] }

 [dev-dependencies]
-syslog_rfc5424 = "0.10"
+syslog_rfc5424 = "0.9"
 tracing-subscriber = { version = "0.3.22", default-features = false, features = ["registry"] }
```
If you give me permission to push to your branch I can do so, or, since the patch is so simple, you can just apply it yourself & update the PR-- then I'll merge it.

cbeck88 committed:
fix build in edition 2021

cbeck88 committed:
use a permalink for the NormalizeMetadata example

cbeck88 commented:
done, thank you :)

sp1ff commented:
I merged this PR by hand; not sure why it's still showing as "open" (perhaps because I locally squashed a bunch of small commits before merging?). Anyway.

Thanks again for this PR. I'm going to write some documentation I've been meaning to do for a while this afternoon, then cut a new release on crates.io.

cbeck88 commented:
thank you!

sp1ff commented:
It's up on crates.io.

BTW, I tried to visit your personal site (https://chrisbeck.co/) and it's not responding for me.
